### PR TITLE
Initializes Selected cases instance For Multi Select Case Lists 

### PR DESCRIPTION
**Pull request by shubham1g5**
https://github.com/dimagi/formplayer/pull/1112

### CODE REVIEW

File: src/main/java/org/commcare/data/xml/VirtualInstances.java [hunk 1]
```diff
@@ -24,4 +28,15 @@ public static VirtualDataInstance buildSearchInputInstance(Map<String, String> u
...
+
+    public static VirtualDataInstance buildSelectedValuesInstance(
+            ExternalDataInstance instance, String[] selectedValues) {
+        List<SimpleNode> nodes = new ArrayList<>();
+        for (String selectedValue : selectedValues) {
+            nodes.add(SimpleNode.textNode(SELCTED_CASES_INSTANCE_NODE_NAME, selectedValue));
+        }
+        TreeElement root = TreeBuilder.buildTree(instance.getInstanceId(), SELCTED_CASES_INSTANCE_ROOT_NAME,
+                nodes);
+        return new VirtualDataInstance(SEARCH_INSTANCE_ID, root);
```
Comment (snopoke) [inline src/main/java/org/commcare/data/xml/VirtualInstances.java:40]: This needs a different instance ID
```suggestion
        return new VirtualDataInstance("search-input", root);
```

Comment (shubham1g5) [reply src/main/java/org/commcare/data/xml/VirtualInstances.java]: thanks for catching, [Corrects instanceid](https://github.com/dimagi/commcare-core/pull/1084/commits/6ea2a14df19497d67239586b8992aee0fa4a6d39)

File: src/main/java/org/commcare/core/process/CommCareInstanceInitializer.java [hunk 2]
```diff
@@ -89,10 +90,27 @@ public InstanceRoot generateRoot(ExternalDataInstance instance) {
             return setupRemoteData(instance);
         } else if (ref.contains("migration")) {
             return setupMigrationData(instance);
+        } else if (ref.contains("selected_cases")) {
+            return setupSelectedCases(instance);
         }
         return ConcreteInstanceRoot.NULL;
     }
 
+    protected InstanceRoot setupSelectedCases(ExternalDataInstance instance) {
```
Comment (snopoke) [inline src/main/java/org/commcare/core/process/CommCareInstanceInitializer.java:99]: Would be good to add javadoc with a note that this should be implemented by subclasses.

File: src/cli/java/org/commcare/util/screen/MultiSelectEntityScreen.java [hunk 3]
```diff
@@ -87,6 +87,17 @@ protected void updateSession(CommCareSession session) {
         }
     }
 
+    @Override
+    public boolean referencesContainStep(String stepValue) {
+        try {
+            String[] cachedSelection = entitiesSelectionCache.read(stepValue);
+            return cachedSelection != null;
+        } catch (SQLException throwables) {
+            throwables.printStackTrace();
```
Comment (snopoke) [inline src/cli/java/org/commcare/util/screen/MultiSelectEntityScreen.java:96]: Does this end up being logged anywhere? I would have though that using a logger would be better.

File: src/cli/java/org/commcare/util/screen/MultiSelectEntityScreen.java [hunk 4]
```diff
@@ -87,6 +87,15 @@ protected void updateSession(CommCareSession session) {
         }
     }
 
+    @Override
+    public boolean referencesContainStep(String stepValue) {
```
Comment (snopoke) [inline src/cli/java/org/commcare/util/screen/MultiSelectEntityScreen.java:91]: Where will this get used?

Comment (shubham1g5) [reply src/cli/java/org/commcare/util/screen/MultiSelectEntityScreen.java:91]: This gets called when [we rebuild the menu session from session frame](https://github.com/dimagi/formplayer/blob/master/src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java#L90) after EOF navigation or case claim to establish the validity of the datum in the session frame.

**Review by snopoke — APPROVED** [review]

**Pull request merged**